### PR TITLE
Multiple versions

### PR DESCRIPTION
**Pull request by WanderingStar**
This adds support for multiple versions of Articles (differentiated by the `updated` column).

An instance of an Article can `get_updated_version` (which returns the latest version of that Article), and `create_new_version` (which creates a new version with a specified `status`, changing the Article in place... that is no need to say `article2 = article1.create_new_version(status)`).

The Article class has class methods to get the latest Article for a given url/url_id, and to create a SELECT statement that uses a window function to only select the most recent version of any articles that match subsequent filters.

### CODE REVIEW

File: source/python/idetect/model.py [hunk 1]
```diff
@@ -65,35 +65,76 @@ class Article(Base):
...
+        block until the transaction is committed or rolled back.
         """
-        Atomically Update the status of this Article from to new_status.
-        If something changed the status of this article since it was loaded, raise.
+        query = session.query(cls)
+        if url_id is not None:
+            query = query.filter(Article.url_id == url_id)
+        if url is not None:
+            query = query.filter(Article.url == url)
+        # if neither url_id nor url is provided, this returns the most recent article overall
```
Comment (simonb83) [inline source/python/idetect/model.py:98]: Is there a scenario where both `url` and `url_id` are provided?

File: source/python/idetect/model.py [hunk 2]
```diff
@@ -65,35 +65,76 @@ class Article(Base):
...
+            if not session:
+                raise RuntimeError("Object has not been persisted in a session.")
+
+            # lock to prevent simultaneous updates
+            latest = Article.get_latest_version(session, self.url_id, lock=True)
+            if latest.id != self.id:
+                raise NotLatestException(self, latest)
+
+            make_transient(self)
+            self.id = None
```
Comment (simonb83) [inline source/python/idetect/model.py]: Does `make_transient` also copy over the content mapping?

File: source/python/idetect/worker.py [hunk 3]
```diff
@@ -45,29 +46,35 @@ def work(self):
...
+                    # ... pick the first (oldest)
+                    article = Article.select_latest_version(self.session) \
+                        .filter(Article.status == self.status) \
+                        .order_by(Article.updated) \
+                        .first()
                     if article is None:
                         return False  # no work to be done
-                    article.update_status(self.working_status)
+                    article.create_new_version(self.working_status)
                     self.session.commit()
```
Comment (simonb83) [inline source/python/idetect/worker.py:60]: So are we creating a new version of an article every time we do something?

File: source/python/idetect/model.py (hunk 1 above)
Comment (WanderingStar) [reply source/python/idetect/model.py:98]: There probably isn't, in practice, but if both are there only a row that matches both will be returned.

File: source/python/idetect/model.py (hunk 2 above)
Comment (WanderingStar) [reply source/python/idetect/model.py]: Nope. Good catch. We'll need to add those mappings.

File: source/python/idetect/worker.py (hunk 3 above)
Comment (WanderingStar) [reply source/python/idetect/worker.py:60]: Right. To create an edit history for the Article, we're updating the version whenever it's changed.

File: source/python/idetect/model.py (hunk 2 above)
Comment (simonb83) [reply source/python/idetect/model.py]: Cool. We'll need to figure out all the different scenarios that should/shouldn't copy the mappings.
I think we had also said that in some scenarios we would copy the extracted reports too.

**Review by simonb83 — APPROVED** [review]
I think this is great for the time being

**Pull request merged**